tests/extras/test_ftrl.py

```diff
@@ -606,6 +606,26 @@ def test_ftrl_fit_predict_from_setters():
     assert_equals(target1, target2)
 
 
+def test_ftrl_fit_predict_view():
+    ft = Ftrl(d=100)
+    df_train = dt.Frame([random.random() for _ in range(ft.d)])
+    df_target = dt.Frame([bool(random.getrandbits(1)) for _ in range(ft.d)])
+    rows = range(ft.d//2, ft.d)
+
+    ft.fit(df_train[rows,:], df_target[rows,:])
+    predictions = ft.predict(df_train[rows,:])
+    model = dt.Frame(ft.model.to_dict())
```

Comment:
@st-pasha, this is something I don't understand. If I just do `model = dt.Frame(ft.model)`, then `model` is still a reference to `ft.model`. Why is it so?

Reply:
`dt.Frame(DT)` creates a shallow copy of `DT`. This is equivalent to `DT.copy()`. What makes you think it doesn't work?

Reply:
Ah, I see. I don't say it doesn't work. I just didn't expect `dt.Frame(DT)` to return a shallow copy. What is the best way to make a deep copy then?

Reply:
I believe `copy.deepcopy()` just pickles and then unpickles the object. The more important question is why do you need a deep copy? The datatable should work correctly even with shallow copies.

Reply:
Because I need a copy of the model here, otherwise if I simply do

```
model = dt.Frame(ft.model)
ft.reset
```

`model` is reset also

Reply:
I'll need to review that code, but it seems like there is a bug there. I would expect that simple `model = ft.model; ft.reset()` (without any copying) would preserve the variable `model`.

Reply:
If I just do (for `d=4`, i.e. model size is 4x2)

```
    model = ft.model
    print("model: ", model.to_list())
    ft.reset()
    print("model after reset: ", model.to_list())
```

this is what I get

```
model:  [[-0.5, 0.0, 0.0, 0.5], [0.25, 0.0, 0.0, 0.25]]
model after reset:  [[0.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0, 0.0]]
```

To me it looks consistent, because the model is returned as a shallow copy from FTRL: `return dt_model->copy();`. If you think there is a bug in my code, or somewhere else, please let me know. Thanks!

Reply:
I've looked into this a little, and I think there could be a problem with pointers `z` and `n`. Even though they are acquired as `column->data_w()` pointers, they are kept for an arbitrary long period of time. During that time the pointers may become invalid (in particular, they may become invalid for writing if you create a copy of the Frame). It might be better to re-acquire these pointers at the beginning of each python call that needs them.

Reply:
Are you sure? Here's what I'm getting:
```
>>> import random
>>> import datatable as dt
>>> from datatable.models import Ftrl
>>> ft = Ftrl(d=4)
>>> df_train = dt.Frame([random.random() for _ in range(ft.d)])
>>> df_target = dt.Frame([bool(random.getrandbits(1)) for _ in range(ft.d)])
>>> ft.fit(df_train, df_target)
>>> model = ft.model
>>> model.to_list()
[[0.9553244374605763, 0.0, 0.0, 0.0], [1.0001815001713965, 0.0, 0.0, 0.0]]
>>> ft.reset()
>>> model.to_list()
[[0.9553244374605763, 0.0, 0.0, 0.0], [1.0001815001713965, 0.0, 0.0, 0.0]]
>>> print(ft.model)
None
```

Reply:
I've recompiled datatable from scratch and it seems like this problem has been fixed. I will update tests. Thanks!

Actually there was also another problem when I just did `df_train_range = dt.Frame(df_train[rows,:])`, it kind of inherited rowindex and `df_train_range` was still a view. That is why I had to do `df_train_range = dt.Frame(df_train[rows,:].to_list())`. Is it something we expect? If yes, then what is the best way to create a real frame based on the view?

Reply:
for any frame you can write `frame.materialize()`, and it will be converted from "view" into a "real" data frame, in-place.